### ISSUE TITLE
Are the default hyperparameter settings optimal?

### ISSUE THREAD

**Issue by wending0417** [zjunlp/EasyEdit#121]
Hello! Thanks to your great contributions!
I intend to utilize MEND to edit models such as gpt2-xl, llama2-7b. Are the default hyperparameter settings in this project optimal? Can I directly train hyper network models and edit language models according to the configuration files?

**Comment by pengzju** [comment]
Model editing is less sensitive to hyperparameters, so you can train directly with the default configuration.

**Comment by pengzju** [comment]
For example, if you want to edit `gpt2-xl` using `MEND`, you can use the configuration file [gpt2.xl.yaml](https://github.com/zjunlp/EasyEdit/blob/main/hparams/TRAINING/MEND/gpt2-xl.yaml) to train.
Once the weights for the training are loaded, edit them using the following default configuration file:
https://github.com/zjunlp/EasyEdit/blob/main/hparams/MEND/gpt2-xl.yaml

**Comment by wending0417** [comment]
Thank you for your reply. 
I have tried in this way using default configurations. I trained models for MEND to edit gpt2-xl、gpt-j-6b、llama2-7b on COUNTERFACT dataset(counterfact-train.json and counterfact-val.json) and then edited these models in some random cases (number = 45) of counterfact-edit.json. The results of gpt-j-6b is regular, but the results of gpt2-xl is very poor (reliability = 0.0) and the results of llama2-7b is not good enough (reliability=0.45).

**Comment by pengzju** [comment]
I guess the reason is underfitting. You can check the early_stop_key field to see if the corresponding value is normal.
Another way to prevent underfitting is to save the weights at the last step of the training. You might consider adding `self.save_state(val_info)` [here](https://github.com/zjunlp/EasyEdit/blob/main/easyeditor/trainer/BaseTrainer.py#L184).

**Comment by wending0417** [comment]
I also think this might be the reason. But I didn’t  modify the early stop configurations
`early_stop_patience: 20000
early_stop_key: "loss/total_edit_val"`
So I am confused about if it is the optimal configuration.

**Comment by pengzju** [comment]
In theory, yes. The problem is that there is no optimal weight saved, and the hyperparameter just needs to be set as the default.

**Comment by wending0417** [comment]
Thank you for your guidance. I will try the method you suggested and see if it works.

**Comment by wending0417** [comment]
I have tried the way you suggested:
> Another way to prevent underfitting is to save the weights at the last step of the training. You might consider adding self.save_state(val_info) [here](https://github.com/zjunlp/EasyEdit/blob/main/easyeditor/trainer/BaseTrainer.py#L184).
But it didn’t work. I tried to edit llama2-7b with trained model, the reliability metric is about 0.3. 
Though I saved the weights at the last step of the training, I checked the training process. The process was ended because "No decrease in edit/acc_val for 30000 steps". It seemed that the early stop setting worked.
Training results is:
loss/edit_val       :  4.92601
loss/loc_val        :  0.05808
edit/acc_val        :  0.45646
edit/log_prob_val   : -4.92601
edit/prob_val       :  0.43176
acc/pre_val         :  0.00000
acc/post_val        :  0.00279
nll/pre_val         :  17.01758
perplexity/pre_val  :  24583294.00000
nll/post_val        :  17.45086
perplexity/post_val :  37915188.00000
n_tokens/pre_val    :  12.60417
n_tokens/post_val   :  12.60417
time/edit_val       :  0.93520
loss/total_val      :  0.55068
loss/total_edit_val :  0.55068
memory/alloc_max_val:  0.00000
memory/res_max_val  :  0.00000
eval_time/elapsed   :  657.35785
eval_time/average   :  0.34255

**Comment by pengzju** [comment]
- Sorry for the inconvenience, make sure to use a single GPU when you train `mend`❗️.
    - You should set `model_parallel` to `False` in the configuration file.


Reason:
- During our implementation, mend lost gradients through parallel training of multiple GPUs, and this bug has not been fixed yet.
    - https://github.com/zjunlp/EasyEdit/blob/main/easyeditor/trainer/EditTrainer.py#L69 (so we ignore backpropagation)
    - Make sure `allow_unused` is `False`

**Comment by zxlzr** [comment]
Hi, do you have any further question?

**Comment by wending0417** [comment]
Actually all the training processes were on a single GPU (model_parallel was set to False). But the performance was poor.
I trained MEND for gpt2-xl, the trainning process took 100000 steps which is the max_iters, and I saved the weights at the last step according to your suggestion(`self.save_state(val_info)`)
but I test the hyper network on 50 random cases, only one succeeded in editing, and others all failed.

**Comment by wending0417** [comment]
I found you have provided trained MEND models for gpt2-xl、gptj、llama2-7b in https://github.com/zjunlp/EasyEdit/issues/66. But the models were trained on zsRE. May you provided the models trained on COUNTERFACT? Because I can't reproduce the results with my own trained models.
It would be much appreciated!

**Comment by pengzju** [comment]
- Sorry, we didn't use mend to train multiple models (the three mentioned above) in our experiment, so I don't have checkpoints here.
- Are you sure that the edit set and the train set are in the same domain?
- Perhaps you can provide the logs during the training stage, such as the changes of `acc` or `loss`, etc. At present, I have no way to know the reason why you cannot reproduce

**Comment by wending0417** [comment]
Yes, the train set is "counterfact-train.json", validate set is "counterfact-val.json", test set is a subset of "counterfact-edit.json", all the files are acquired from your repo.
I will send you the training logs. Thank you for your help!

**Comment by wending0417** [comment]
Hi, sorry for bothering you. Did you receive the training log?

**Comment by pengzju** [comment]
Judging from the results of the `edit/acc_train` metric, your experiment results are correct. I have the following two suggestions

- Set `val_batch_size` to `1` (5 in your log file), which may reproduce the result on val set.
- Directly load the trained additional network and edit it on `XXX-edit.json` to observe the experimental results

**Comment by wending0417** [comment]
Sorry, I'm a little confused. 

> your experiment results are correct.

Does that mean the trained additional network is correct?

> Directly load the trained additional network and edit it on XXX-edit.json to observe the experimental results

Is this step just to increase the size of the test dataset, or does it have other significance?
Thank you!

**Comment by pengzju** [comment]
> Does that mean the trained additional network is correct?


Yes, because I observed that `edit/train_acc` is as expected.

> Is this step just to increase the size of the test dataset, or does it have other significance?


As above, since there is no problem with the trained network, I recommend you to edit directly on `counterfact-edit.json` to test accuracy

**Comment by wending0417** [comment]
Ok, I will try your suggestion. And I will reply if the results are reproduced. Thank you!

**Comment by wending0417** [comment]
Hi, I tried to directly edit on `counterfact-edit.json` and calculate the accuracy:
MEND + gpt2-xl + counterfact = 4.79%
MEND+ llama2-7b + counterfact = 24.34%
Is this acceptable?

**Comment by wending0417** [comment]
Actually, I have set `keep_original_weight` to `True`.

**Comment by wending0417** [comment]
May I inquire if you have tried the combination of MEND + gpt2-xl/llama2-7b + COUNTERFACT, and are the results normal?

**Comment by pengzju** [comment]
>  May I inquire if you have tried the combination of MEND + gpt2-xl/llama2-7b + COUNTERFACT, and are the results normal?


- I can definitely tell you that the combination of `mend+ LLAMA7b + counterfact` is able to reproduce the results in the original paper, which is some of our experience. 
- However, I have not tried the experiment of editing `gpt2`. I guess it may be that the convolutional network in gpt2 has a certain particularity. But according to your indicators, everything looks fine. I suggest you check the **input and output** to see if there is a **certain pattern of errors**

**Comment by wending0417** [comment]
I checked the log file of gpt2,
`loss/edit_val:10.79611` and `edit/acc_val:0.02678` is far lower than others, maybe the trained hyper network is not correct.

**Comment by pengzju** [comment]
> I checked the log file of gpt2, `loss/edit_val:10.79611` and `edit/acc_val:0.02678` is far lower than others, maybe the trained hyper network is not correct.

The results from acc are incorrect.

Thank you for your feedback. I will keep this issue open and try the combination of `mend+gpt2-xl+counterfact` as soon as possible.

**Comment by wending0417** [comment]
Thank you! Looking forward to your result!

**Comment by wending0417** [comment]
Hi~, is there any progress？
I have tried to use the MEND models for gpt2-xl you provided (https://github.com/zjunlp/EasyEdit/issues/66) to edit on ZsRE. But the results are really poor, there is no case achieve 1.0 for the metric `rewrite_acc` from 100 test cases.

**Comment by littlefive5** [comment]
Sorry for the slow response.
If you are urgent, maybe you can try the checkpoint for gpt2-xl provided by the original paper? https://rome.baulab.info/data/weights/
Peng is busy currently; I will try to trace this issue and figure out the problem.
Will let you know soon.

**Comment by pengzju** [comment]
Hello. I hope everything is going well.
I checked the training process and the problem was that the weights were not well preserved. Currently I have updated the EasyEdit code, you can update your own native library to repeat the MEND+gpt2-xl+CounterFact experiment.
In my latest experiment, I was able to replicate the 1.0 editing results. As shown in the following figure:
<img width="1032" alt="image" src="https://github.com/zjunlp/EasyEdit/assets/63391142/dce02cbb-a01e-4481-b519-1e8a475d8cf4">


You can also use my weights, link at: https://drive.google.com/drive/folders/1SLxf21imI9tgiF8mM-XKfdGUCdaZt99J

**Comment by wending0417** [comment]
Thank you very much for resolving this issue. If the solution works out fine, I will close this issue. However, we do not have access to the link you provided. Thank you once again！

**Comment by pengzju** [comment]
> Thank you very much for resolving this issue. If the solution works out fine, I will close this issue. However, we do not have access to the link you provided. Thank you once again！

I'm very sorry, I have opened the permissions of google drive, and now you can access it normally

**Comment by wending0417** [comment]
Hi, I have tried to use your weights. Although I replicated cases with success rates of 1.0, the success rate is only 12% across 100 test samples. May I ask what your success rate is?

**Comment by wending0417** [comment]
Actually, I have set keep_original_weight to True. Maybe there is other bug in my code, I will try to find it.

**Comment by wending0417** [comment]
Hi, after updating my local code to match the current repository version and using your weights, I was able to replicate the result. Thank you for resolving this issue.
